# Lab 9. My second data science project

## In this lab you will apply the concepts and techniques developed in previous sessions. You are expected to work on the assignments on your own DURING the duration of the lab.
## You are expected to submit the code developed as well as a brief description of your findings and insights for each assigment AT THE END of the lab. 

## This lab represents 20% of the total marking of the module¶

In [3]:
!pip install --upgrade pandas

  Using cached pandas-1.2.3-cp38-cp38-win_amd64.whl (9.3 MB)


ERROR: Could not install packages due to an OSError: [WinError 5] Acceso denegado: 'c:\\programdata\\anaconda3\\lib\\site-packages\\pandas-1.1.3.dist-info\\direct_url.json'
Consider using the `--user` option or check the permissions.



  Attempting uninstall: pandas
    Found existing installation: pandas 1.1.3
    Uninstalling pandas-1.1.3:


In [4]:
!pip install py-openaq

In [5]:
import pandas as pd
import seaborn as sns

In [6]:
import openaq

In [7]:
pd.__version__

'1.1.3'

In [8]:
api = openaq.OpenAQ()


In [9]:
London = api.measurements(city='London', limit=5000, df=True)
Paris = api.measurements(city='Paris', limit=5000, df=True)
Madrid = api.measurements(city='Madrid', limit=5000, df=True)
Berlin = api.measurements(city='Berlin', limit=5000, df=True)
Rome = api.measurements(city='Roma', limit=5000, df=True)


C:\ProgramData\Anaconda3\lib\site-packages\openaq\decorators.py:57: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  data = pd.io.json.json_normalize(resp)


In [10]:
Cities=pd.concat([London,Paris,Madrid,Berlin,Rome])

In [11]:
Cities

location parameter  value  \
date.local                                                                
2021-03-24 02:00:00  London Haringey Priory Park South        o3   71.0   
2021-03-24 02:00:00                    Camden Kerbside      pm25   17.0   
2021-03-24 02:00:00                  Haringey Roadside       no2   24.0   
2021-03-24 02:00:00                    Camden Kerbside       no2   16.0   
2021-03-24 02:00:00               London N. Kensington      pm10   28.0   
...                                                ...       ...    ...   
2021-03-18 06:00:00                            IT2171A       no2   34.0   
2021-03-18 06:00:00                            IT0953A        o3    2.0   
2021-03-18 06:00:00                            IT2012A       so2    1.8   
2021-03-18 06:00:00                            IT2012A        o3   31.0   
2021-03-18 06:00:00                            IT0885A       no2   15.0   

                                       unit country    city  \
date.local                                                    
2021-03-24 02:00:00  b'\xc2\xb5g/m\xc2\xb3'      GB  London   
2021-03-24 02:00:00  b'\xc2\xb5g/m\xc2\xb3'      GB  London   
2021-03-24 02:00:00  b'\xc2\xb5g/m\xc2\xb3'      GB  London   
2021-03-24 02:00:00  b'\xc2\xb5g/m\xc2\xb3'      GB  London   
2021-03-24 02:00:00  b'\xc2\xb5g/m\xc2\xb3'      GB  London   
...                                     ...     ...     ...   
2021-03-18 06:00:00  b'\xc2\xb5g/m\xc2\xb3'      IT    Roma   
2021-03-18 06:00:00  b'\xc2\xb5g/m\xc2\xb3'      IT    Roma   
2021-03-18 06:00:00  b'\xc2\xb5g/m\xc2\xb3'      IT    Roma   
2021-03-18 06:00:00  b'\xc2\xb5g/m\xc2\xb3'      IT    Roma   
2021-03-18 06:00:00  b'\xc2\xb5g/m\xc2\xb3'      IT    Roma   

                                     date.utc  coordinates.latitude  \
date.local                                                            
2021-03-24 02:00:00 2021-03-24 02:00:00+00:00             51.584128   
2021-03-24 02:00:00 2021-03-24 02:00:00+00:00             51.544210   
2021-03-24 02:00:00 2021-03-24 02:00:00+00:00             51.599300   
2021-03-24 02:00:00 2021-03-24 02:00:00+00:00             51.544210   
2021-03-24 02:00:00 2021-03-24 02:00:00+00:00             51.521050   
...                                       ...                   ...   
2021-03-18 06:00:00 2021-03-18 05:00:00+00:00             42.094170   
2021-03-18 06:00:00 2021-03-18 05:00:00+00:00             41.932778   
2021-03-18 06:00:00 2021-03-18 05:00:00+00:00             41.875556   
2021-03-18 06:00:00 2021-03-18 05:00:00+00:00             41.875556   
2021-03-18 06:00:00 2021-03-18 05:00:00+00:00             42.091667   

                     coordinates.longitude  
date.local                                  
2021-03-24 02:00:00              -0.125254  
2021-03-24 02:00:00              -0.175269  
2021-03-24 02:00:00              -0.068218  
2021-03-24 02:00:00              -0.175269  
2021-03-24 02:00:00              -0.213492  
...                                    ...  
2021-03-18 06:00:00              11.795560  
2021-03-18 06:00:00              12.506944  
2021-03-18 06:00:00              12.345000  
2021-03-18 06:00:00              12.345000  
2021-03-18 06:00:00              11.802500  

[25979 rows x 9 columns]

In [12]:
Cities.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 25979 entries, 2021-03-24 02:00:00 to 2021-03-18 06:00:00
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype              
---  ------                 --------------  -----              
 0   location               25979 non-null  object             
 1   parameter              25979 non-null  object             
 2   value                  25979 non-null  float64            
 3   unit                   25979 non-null  object             
 4   country                25979 non-null  object             
 5   city                   25979 non-null  object             
 6   date.utc               25979 non-null  datetime64[ns, UTC]
 7   coordinates.latitude   25979 non-null  float64            
 8   coordinates.longitude  25979 non-null  float64            
dtypes: datetime64[ns, UTC](1), float64(3), object(5)
memory usage: 2.0+ MB


In [13]:
Cities.groupby('parameter').count()

location  value   unit  country   city  date.utc  \
parameter                                                     
co             1272   1272   1272     1272   1272      1272   
no2           10859  10859  10859    10859  10859     10859   
o3             5803   5803   5803     5803   5803      5803   
pm10           5115   5115   5115     5115   5115      5115   
pm25           1951   1951   1951     1951   1951      1951   
so2             979    979    979      979    979       979   

           coordinates.latitude  coordinates.longitude  
parameter                                               
co                         1272                   1272  
no2                       10859                  10859  
o3                         5803                   5803  
pm10                       5115                   5115  
pm25                       1951                   1951  
so2                         979                    979

### Assignment 9.1. Compute the average values for each parameter and city

#### Tip: consider groupby command

### Asignment 9.2. Which city is the most polluted in terms of co, no2 and pm10?

### Tip. consider: https://seaborn.pydata.org/generated/seaborn.lineplot.html

### Asignment 9.3. Is there any relationship between levels of pollution ?

### Tip: 

1. Consider the pivot_table operation, refer to Lab 6.
2. Consider  https://seaborn.pydata.org/examples/scatterplot_matrix.html 